<h1>ACTIVITY1 - PALMER PENGUINS DATA ANALYTICS PROJECT</h1>

<h3>Analyst: Joshua Sumaginsing</h3>

In [82]:
import numpy as np
import pandas as pd
import warnings
import matplotlib.pyplot as plt
warnings.filterwarnings("ignore")

In [83]:
penguins= pd.read_csv("datasets\\penguins.csv")

In [84]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

<h1>Descriptive Analytics</h1>

In [85]:
penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

In [86]:
penguins.describe().T

count         mean         std     min       25%      50%  \
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                      75%     max  
bill_length_mm       48.5    59.6  
bill_depth_mm        18.7    21.5  
flipper_length_mm   213.0   231.0  
body_mass_g        4750.0  6300.0  
year               2009.0  2009.0

<h1>Filtering Data</h1>

In [87]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

In [88]:
penguins.value_counts()

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  sex     year
Adelie   Biscoe  34.5            18.1           187.0              2900.0       female  2008    1
                 35.0            17.9           190.0              3450.0       female  2008    1
                                                192.0              3725.0       female  2009    1
                 35.3            18.9           187.0              3800.0       female  2007    1
                 35.5            16.2           195.0              3350.0       female  2008    1
                                                                                               ..
Gentoo   Biscoe  53.4            15.8           219.0              5500.0       male    2009    1
                 54.3            15.7           231.0              5650.0       male    2008    1
                 55.1            16.0           230.0              5850.0       male    2009    1
                 55.9    

In [89]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

In [90]:
penguins['species'] == 'Adelie'

0       True
1       True
2       True
3       True
4       True
       ...  
339    False
340    False
341    False
342    False
343    False
Name: species, Length: 344, dtype: bool

In [91]:
penguins[penguins['species'] == 'Adelie']

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
147  Adelie      Dream            36.6           18.4              184.0   
148  Adelie      Dream            36.0           17.8              195.0   
149  Adelie      Dream            37.8           18.1              193.0   
150  Adelie      Dream            36.0           17.1              187.0   
151  Adelie      Dream            41.5           18.5              201.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
147       3475.0  female  2009  
148       3450.0  female  2009  
149       3750.0    male  2009  
150       3700.0  female  2009  
151       4000.0    male  2009  

[152 rows x 8 columns]

In [92]:
penguins["species"]

0         Adelie
1         Adelie
2         Adelie
3         Adelie
4         Adelie
         ...    
339    Chinstrap
340    Chinstrap
341    Chinstrap
342    Chinstrap
343    Chinstrap
Name: species, Length: 344, dtype: object

In [93]:
ad_df = penguins[penguins['species'] == 'Adelie']

In [94]:
ad_df

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
147  Adelie      Dream            36.6           18.4              184.0   
148  Adelie      Dream            36.0           17.8              195.0   
149  Adelie      Dream            37.8           18.1              193.0   
150  Adelie      Dream            36.0           17.1              187.0   
151  Adelie      Dream            41.5           18.5              201.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
147       3475.0  female  2009  
148       3450.0  female  2009  
149       3750.0    male  2009  
150       3700.0  female  2009  
151       4000.0    male  2009  

[152 rows x 8 columns]

In [95]:
gen_df = penguins[penguins['species'] == 'Gentoo']

In [96]:
gen_df

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
152  Gentoo  Biscoe            46.1           13.2              211.0   
153  Gentoo  Biscoe            50.0           16.3              230.0   
154  Gentoo  Biscoe            48.7           14.1              210.0   
155  Gentoo  Biscoe            50.0           15.2              218.0   
156  Gentoo  Biscoe            47.6           14.5              215.0   
..      ...     ...             ...            ...                ...   
271  Gentoo  Biscoe             NaN            NaN                NaN   
272  Gentoo  Biscoe            46.8           14.3              215.0   
273  Gentoo  Biscoe            50.4           15.7              222.0   
274  Gentoo  Biscoe            45.2           14.8              212.0   
275  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  year  
152       4500.0  female  2007  
153       5700.0    male  2007  
154       4450.0  female  2007  
155       5700.0    male  2007  
156       5400.0    male  2007  
..           ...     ...   ...  
271          NaN     NaN  2009  
272       4850.0  female  2009  
273       5750.0    male  2009  
274       5200.0  female  2009  
275       5400.0    male  2009  

[124 rows x 8 columns]

In [97]:
chin_df = penguins[penguins['species'] == 'Chinstrap']

In [98]:
chin_df

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
276  Chinstrap  Dream            46.5           17.9              192.0   
277  Chinstrap  Dream            50.0           19.5              196.0   
278  Chinstrap  Dream            51.3           19.2              193.0   
279  Chinstrap  Dream            45.4           18.7              188.0   
280  Chinstrap  Dream            52.7           19.8              197.0   
..         ...    ...             ...            ...                ...   
339  Chinstrap  Dream            55.8           19.8              207.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
276       3500.0  female  2007  
277       3900.0    male  2007  
278       3650.0    male  2007  
279       3525.0  female  2007  
280       3725.0    male  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[68 rows x 8 columns]

In [99]:
ad_df["body_mass_g"].mean()

np.float64(3700.662251655629)

In [100]:
gen_df["body_mass_g"].mean()

np.float64(5076.016260162602)

In [101]:
chin_df["body_mass_g"].mean()

np.float64(3733.0882352941176)

<h1>Conditional operators</h1>

In [102]:
ad_male = penguins[(penguins['species'] == 'Adelie') & (penguins['sex'] == 'male')]

In [103]:
ad_male

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
5    Adelie  Torgersen            39.3           20.6              190.0   
7    Adelie  Torgersen            39.2           19.6              195.0   
13   Adelie  Torgersen            38.6           21.2              191.0   
14   Adelie  Torgersen            34.6           21.1              198.0   
..      ...        ...             ...            ...                ...   
143  Adelie      Dream            40.7           17.0              190.0   
145  Adelie      Dream            39.0           18.7              185.0   
146  Adelie      Dream            39.2           18.6              190.0   
149  Adelie      Dream            37.8           18.1              193.0   
151  Adelie      Dream            41.5           18.5              201.0   

     body_mass_g   sex  year  
0         3750.0  male  2007  
5         3650.0  male  2007  
7         4675.0  male  2007  
13        3800.0  male  2007  
14        4400.0  male  2007  
..           ...   ...   ...  
143       3725.0  male  2009  
145       3650.0  male  2009  
146       4250.0  male  2009  
149       3750.0  male  2009  
151       4000.0  male  2009  

[73 rows x 8 columns]

<h1>Cleaning the Data Frames</h1>

In [104]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h1>What to do with NULL VALUES</h1>

In [105]:
penguins['bill_length_mm'] = penguins['bill_length_mm'].fillna(penguins['bill_length_mm'].mean())

In [106]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [107]:
penguins['bill_depth_mm'] = penguins['bill_depth_mm'].fillna(penguins['bill_depth_mm'].mean())

In [108]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [109]:
penguins['flipper_length_mm'] = penguins['flipper_length_mm'].fillna(penguins['flipper_length_mm'].mean())

In [110]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     0
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [111]:
penguins['body_mass_g'] = penguins['body_mass_g'].fillna(penguins['body_mass_g'].mean())

In [112]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     0
body_mass_g           0
sex                  11
year                  0
dtype: int64

In [113]:
penguins['sex'] = penguins['sex'].fillna(penguins['sex'].mode()[0])


In [114]:
penguins.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

<h1>CHANGING THE DATA TYPES</h1>

In [115]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     344 non-null    float64
 3   bill_depth_mm      344 non-null    float64
 4   flipper_length_mm  344 non-null    float64
 5   body_mass_g        344 non-null    float64
 6   sex                344 non-null    object 
 7   year               344 non-null    int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 21.6+ KB


In [116]:
penguins['species'] = penguins['species'].astype('category')
penguins['island'] = penguins['island'].astype('category')
penguins['sex'] = penguins['sex'].astype('category')
penguins['year'] = penguins['year'].astype('category')

In [117]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    category
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      344 non-null    float64 
 4   flipper_length_mm  344 non-null    float64 
 5   body_mass_g        344 non-null    float64 
 6   sex                344 non-null    category
 7   year               344 non-null    category
dtypes: category(4), float64(4)
memory usage: 12.7 KB


In [118]:
penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count      344.000000     344.000000         344.000000   344.000000
mean        43.921930      17.151170         200.915205  4201.754386
std          5.443643       1.969027          14.020657   799.613058
min         32.100000      13.100000         172.000000  2700.000000
25%         39.275000      15.600000         190.000000  3550.000000
50%         44.250000      17.300000         197.000000  4050.000000
75%         48.500000      18.700000         213.000000  4750.000000
max         59.600000      21.500000         231.000000  6300.000000

<h3>1. How many penguins belong to each species?</h3>

In [142]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

Insight: Adelie penguin has 152 , Gentoo has 124 penguins and Chinstrap has 68 penguins

<h3>2. Mean body mass per island <h3/> 

In [119]:
penguins.groupby('island')['body_mass_g'].mean()

island
Biscoe       4712.956871
Dream        3712.903226
Torgersen    3715.899123
Name: body_mass_g, dtype: float64

Insight: Penguins living on Biscoe island are, on average, significantly heavier (4713g) than those on Dream (3713g) and Torgersen (3715g).

<h3>3. Heaviest penguin by sex</h3>

In [120]:
penguins.groupby('sex')['body_mass_g'].max()

sex
female    5200.0
male      6300.0
Name: body_mass_g, dtype: float64

Insight: The heaviest penguin by sex is male penguins are generally heavier than females.

<h3>4. Median bill depth by species</h3>

In [121]:
penguins.groupby('species')['bill_depth_mm'].median()

species
Adelie       18.40
Chinstrap    18.45
Gentoo       15.00
Name: bill_depth_mm, dtype: float64

Insight: Adelie and Chinstrap penguins have deeper bills than Gentoo penguins.

<h3>5.  What is the average flipper length of each species?</h3>

In [131]:
penguins.groupby('species')['flipper_length_mm'].mean()

species
Adelie       190.025758
Chinstrap    195.823529
Gentoo       217.055768
Name: flipper_length_mm, dtype: float64

Insights: On average, Adelie penguins have the shortest flippers, whereas Gentoo penguins have the longest.

<h3>6. Is island correlated with body mass? </h3>

In [123]:
penguins.groupby("island")["body_mass_g"].mean()

island
Biscoe       4712.956871
Dream        3712.903226
Torgersen    3715.899123
Name: body_mass_g, dtype: float64

Insight: Yes. Penguins on Biscoe Island are, on average, about 1 kg heavier
 Penguins from Dream and Torgersen have almost the same lighter average body mass.

<h3>7. Do flipper lengths change with year recorded?</h3>

In [124]:
penguins[['year', 'flipper_length_mm']].corr()

year  flipper_length_mm
year               1.000000           0.168934
flipper_length_mm  0.168934           1.000000

Insight: Flipper length has a weak positive trend over the years.

<h3>8. does the flipper length have effect on bill length</h3>

In [128]:
penguins[["flipper_length_mm","bill_length_mm"]].corr()

flipper_length_mm  bill_length_mm
flipper_length_mm           1.000000        0.656181
bill_length_mm              0.656181        1.000000

Insights: Yes, flipper length have a average effect on bill length

<h3>9. What is the median bill length for each species? <h3/>

In [129]:
penguins.groupby("species")["bill_length_mm"].median()

species
Adelie       38.80
Chinstrap    49.55
Gentoo       47.25
Name: bill_length_mm, dtype: float64

Insight:  Chinstrap penguins have the longest bills on average, while Adelie penguins have the shortest.

<h3>10. What is the lighest Chinstrap Penguins in Dream Island?</h3>

In [140]:
ad_df = penguins[(penguins['species'] == "Chinstrap") & (penguins['island'] == 'Dream')]

In [141]:
ad_df['body_mass_g'].min()

np.float64(2700.0)

Insight: The lightest Chinstrap penguin on Dream Island weighs 2700 g

<h3>11. What is the minimun bill length per island?</h3>

In [147]:
bill_length = penguins.groupby('island')['bill_length_mm'].agg('min')

In [148]:
bill_length

island
Biscoe       34.5
Dream        32.1
Torgersen    33.5
Name: bill_length_mm, dtype: float64

Insight: Penguins on Dream Island have the shortest minimum bill length (32.1 mm) compared to the other islands.

<h3>12. What is the minimum body mass of male and female penguins?.</h3>

In [149]:
penguins.groupby("sex")["body_mass_g"].min()

sex
female    2700.0
male      2975.0
Name: body_mass_g, dtype: float64

Insight: Female penguins can be smaller and lighter than males, consistent with the general pattern of males being the larger sex.